### PR TITLE
Detect static default interface methods in illink

### PR DESCRIPTION
**Pull request by sbomer**
Fixes https://github.com/dotnet/runtime/issues/96516. The issue was that we were skipping the TypeMapInfo logic that would have been needed to discover static default interface implementations.

While investigating this I noticed more problems in this area (I will file bugs) that I plan to fix separately. For example, the current logic only works for instantiated types, so if you replace `struct Derived` in this testcase with a class, it fails.

### CODE REVIEW

**Review by MichalStrehovsky — APPROVED** [review]
Better mark too many methods than too few methods. LGTM.

I double checked that my advice to add the `IsStatic` check in #93662 did not in fact cause this ~~regression~~ bug - the previous `IsNewSlot` check was already filtering statics on its own since Roslyn doesn't mark static virtuals newslot. That PR is scheduled for servicing so it would by annoying if it was a regression.

File: src/tools/illink/test/Mono.Linker.Tests.Cases/Inheritance.Interfaces/DefaultInterfaceMethods/StaticDefaultInterfaceMethodOnStruct.cs [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+		public static void Main ()
+		{
+#if SUPPORTS_DEFAULT_INTERFACE_METHODS
+			Foo<Derived> ();
+#endif
+		}
+
+#if SUPPORTS_DEFAULT_INTERFACE_METHODS
+		[Kept]
+		static void Foo<T> () where T : IBase
```
Comment (vitek-karas) [inline src/tools/illink/test/Mono.Linker.Tests.Cases/Inheritance.Interfaces/DefaultInterfaceMethods/StaticDefaultInterfaceMethodOnStruct.cs:17]:
```suggestion
		static void TestStaticDefaultMethodOnStruct<T> () where T : IBase
```

Or just `Test` - took me a while what it does... and it's just a way to call the method.

File: src/tools/illink/test/Mono.Linker.Tests.Cases/Inheritance.Interfaces/DefaultInterfaceMethods/StaticDefaultInterfaceMethodOnStruct.cs [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+		{
+			T.Foo ();
```
Comment (vitek-karas) [inline src/tools/illink/test/Mono.Linker.Tests.Cases/Inheritance.Interfaces/DefaultInterfaceMethods/StaticDefaultInterfaceMethodOnStruct.cs:19]: Nit: I would avoid `Foo`, `Bar` and similar. Policheck doesn't like those already - and it's not descriptive. Personally, I prefer descriptive names in tests - makes it easier to understand what the test does and typically makes it easier to debug.

**Review by vitek-karas — APPROVED** [review]
Nice - thanks a lot.

**Review by jtschuster — APPROVED** [review]
LGTM, thank you!

**Pull request merged**